walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,149 @@
+        int mandatoryEnd = available.get(indexOfIntersection).end();
+        if (end <= mandatoryEnd) {
+          finalTimeRanges.add(optionalTimeRange);
+        }
+      }
+    }
+
+    return finalTimeRanges.size() != 0 ? finalTimeRanges : available;
+  }
+    
+  public int binarySearchOverlappingTimeRange(ArrayList<TimeRange> mandatory, int startIndex, int endIndex, int startOfOptional) {
+    int startOfMandatory = mandatory.get(startIndex).start();
+
+    if (startIndex == endIndex) {
+      return startOfMandatory <= startOfOptional ? startIndex : -1;  
+    }
+
+    int midIndex = startIndex + (endIndex - startIndex) / 2;
+
+    if (startOfOptional < mandatory.get(midIndex).start()) {
+      return binarySearchOverlappingTimeRange(mandatory, startIndex, midIndex, startOfOptional);
+    }
+
+    int ret = binarySearchOverlappingTimeRange(mandatory, midIndex + 1, endIndex, startOfOptional);
+
+    return ret == -1 ? midIndex : ret;
+  }
+
+  public ArrayList<TimeRange> availableTimesFromSortedEvents(HashSet<String> attendees, ArrayList<Event> events, MeetingRequest request) {
+    ArrayList<TimeRange> availableTimes = new ArrayList<>();
+    ArrayList<Event> sequentialEvents = new ArrayList<>(events);
+
+    for (int i = 0; i < sequentialEvents.size(); ++i) {
+      HashSet<String> attendeesForEvent = new HashSet<String>(sequentialEvents.get(i).getAttendees());
+
+      attendeesForEvent.retainAll(attendees);
+
+      if (attendeesForEvent.size() <= 0) {
+        sequentialEvents.remove(i);
+        --i;
+      }
+    }
+    
+    for (int i = 0; i < sequentialEvents.size() - 1;) {
+      Event eventToCheck = sequentialEvents.get(i);
+      Event nextEvent = sequentialEvents.get(i + 1);
+      
+      int startTimeEvent = eventToCheck.getWhen().start();
+      int endTimeEvent = eventToCheck.getWhen().end();
+      int startTimeNextEvent = nextEvent.getWhen().start();
+      int endTimeNextEvent = nextEvent.getWhen().end();
+          
+      if (endTimeEvent >= startTimeNextEvent) {
+        sequentialEvents.remove(i);
+        sequentialEvents.remove(i);
+        TimeRange mergedTimeRange = TimeRange.fromStartDuration(startTimeEvent, Math.max(endTimeEvent, endTimeNextEvent) - startTimeEvent);
+        Event mergedEvent = new Event("MERGED_EVENT", mergedTimeRange, attendees);
+        sequentialEvents.add(i, mergedEvent);
+      } else {
+        ++i;
+      }
+    }
+      
+    for (int i = 0; i < sequentialEvents.size() - 1; ++i) {
+      
+      Event eventToCheck = sequentialEvents.get(i);
+      Event nextEvent = sequentialEvents.get(i + 1);
+      
+      int endTimeEvent = eventToCheck.getWhen().end();
+      int startTimeNextEvent = nextEvent.getWhen().start();
+      
+      int duration = startTimeNextEvent - endTimeEvent;
+      
+
+      if (duration >= request.getDuration()) {
+        availableTimes.add(TimeRange.fromStartDuration(endTimeEvent, duration));
+      }
+    }
+      
+    return availableTimes;
```

Comment:
General questions:
1) Somewhere in the comments for this class, could you add a plain-English description of the algorithm used to derive meeting times?
2) What is the worst-case runtime complexity (big-O) of this algorithm?